### PR TITLE
Pre-encode headlines, captions, images using tokenizer

### LINKED ISSUE #16: Add script to encode dataset (image tokens)
Using VQGAN encoder, convert all images in the given dataset to image tokens. These image tokens along with tokens from BART encoder (encoded captions/news headline) are fed to BART decoder to train it to generate image tokens given encoded captions/headlines.

We can do this as part of data transform step, however, doing it beforehand will speed up the training.

### PR DESCRIPTION
**Pull request by ArjunKrishna3367**
Closes https://github.com/mahalrs/newsgen/issues/16

Uses tokenizer to encode headlines, captions, and images and saves them to new .json files (headlines_encoded.json and captions_encoded.json)

### CODE REVIEW

File: src/encode_data.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+                    story['image_tokens'] = image_encoding[j].tolist()
+                
+        with open(os.path.join(dataset_dir, f"{file_name[:-5]}_encoding.json"), 'w') as f:
+            json.dump(data, f)
+
+def main():
+    args = parser.parse_args()
+    dataset_dir = args.dataset
+
+    torch.set_grad_enabled(False)
```
Comment (mahalrs) [inline src/encode_data.py]: No need for turning off grads here; tokenizer will handle it

File: src/encode_data.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                        story['headline_tokens'] = headline_encoding[j].tolist()
+                    story['caption_tokens'] = caption_encoding[j].tolist()
+                    story['image_tokens'] = image_encoding[j].tolist()
+                
+        with open(os.path.join(dataset_dir, f"{file_name[:-5]}_encoding.json"), 'w') as f:
+            json.dump(data, f)
+
+def main():
+    args = parser.parse_args()
+    dataset_dir = args.dataset
```
Comment (mahalrs) [inline src/encode_data.py:84]: assert if dataset path exists and if it is a directory

File: src/encode_data.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+from PIL import Image
+from newsgen.tokenizer import NewsgenTokenizer
+
+parser = argparse.ArgumentParser()
+parser.add_argument('--dataset',
+                    default='data/visual_news_mini',
+                    help='Directory containing VisualNews dataset')
+
+
+device = 'cuda' if torch.cuda.is_available() else 'cpu'
```
Comment (mahalrs) [inline src/encode_data.py]: move to main()

File: src/encode_data.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+tok = NewsgenTokenizer('../src/pretrained/exp1.ckpt', device)
```
Comment (mahalrs) [inline src/encode_data.py]: move to main() and pass as argument to encode()

File: src/encode_data.py [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+                    
+                    if 'headline' in story:
+                        headline_batch.append(story['headline'])
+                    caption_batch.append(story['caption'])
+                    image_path = os.path.join(dataset_dir, story['image_path'][2:])
+                    img = Image.open(image_path).convert('RGB')
+                    image_batch.append(img)
+                
+                if len(headline_batch) > 0:
+                    headline_encoding = tok.encode_text_batch(headline_batch)['input_ids']
```
Comment (mahalrs) [inline src/encode_data.py]: Store everything returned, not just 'input_ids'. Attention mask is also needed. Just append everything returned by tokenizer.

File: src/encode_data.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                caption_encoding = tok.encode_text_batch(caption_batch)['input_ids']
```
Comment (mahalrs) [inline src/encode_data.py]: Store everything returned, not just 'input_ids'. Attention mask is also needed. Just append everything returned by tokenizer.

**Comment by mahalrs** [comment]
Related to #20

**Review by mahalrs — CHANGES_REQUESTED** [review]
Let me know when you make requested changes

**Comment by ArjunKrishna3367** [comment]
Fixed most of the requests. Turns out you can't index the batch encoding from the tokenizer directly, so I just pulled out the input_ids and attention_masks and added them as separate fields to the json for each article. Hopefully that works.

File: src/encode_data.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                    
+                    if 'headline' in story:
+                        headline_batch.append(story['headline'])
+                    caption_batch.append(story['caption'])
+                    image_path = os.path.join(dataset_dir, story['image_path'][2:])
+                    img = Image.open(image_path).convert('RGB')
+                    image_batch.append(img)
+                
+                if len(headline_batch) > 0:
+                    headline_encoding = tok.encode_text_batch(headline_batch)
```
Comment (mahalrs) [inline src/encode_data.py]: Are you sure .cpu() is not needed?

`headline_encoding = tok.encode_text_batch(headline_batch).cpu()`

File: src/encode_data.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                    headline_tokens = headline_encoding['input_ids']
+                    headline_attention = headline_encoding['attention_mask']
+                caption_encoding = tok.encode_text_batch(caption_batch)
```
Comment (mahalrs) [inline src/encode_data.py]: .cpu()

`headline_encoding = tok.encode_text_batch(caption_batch).cpu()`

File: src/encode_data.py [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+                caption_tokens = caption_encoding['input_ids']
+                caption_attention = caption_encoding['attention_mask']
+                image_encoding = tok.encode_image_batch(image_batch)
```
Comment (mahalrs) [inline src/encode_data.py]: .cpu()

`image_encoding = tok.encode_image_batch(image_batch).cpu()`

**Review by mahalrs — CHANGES_REQUESTED** [review]
Minor changes

File: src/encode_data.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            json.dump(data, f)
+
+def main():
+    args = parser.parse_args()
+    dataset_dir = args.dataset
+    assert os.path.exists(dataset_dir), f'{dataset_dir} does not exist.'
+    assert os.path.isdir(dataset_dir), f'{dataset_dir} is not a directory.'
+    
+    device = 'cuda' if torch.cuda.is_available() else 'cpu'
+    tok = NewsgenTokenizer('../src/pretrained/exp1.ckpt', device)
```
Comment (mahalrs) [inline src/encode_data.py]: Better to take checkpoint path as command line argument.

File: src/encode_data.py (hunk 7 above)
Comment (ArjunKrishna3367) [reply src/encode_data.py]: This is causing an AttributeError. They're tensors, so I can do `tok.encode_text_batch(headline_batch).to("cpu") `instead, seems to be working.

Comment (mahalrs) [reply src/encode_data.py]: .to('cpu') sounds good

**Review by mahalrs — APPROVED** [review]
/lgtm
